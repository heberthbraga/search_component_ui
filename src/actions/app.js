import types from '../types'

export const goToHome = () => dispatch => ( dispatch({ type: types.REDIRECT_TO_HOME, payload: true }) )