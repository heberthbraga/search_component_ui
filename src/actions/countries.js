import types from '../types'

import CountriesApiService from '../api/countries'

export const fetchCountries = () => dispatch => {
  console.log("Fetching Countries ")

  CountriesApiService.list().then(
    res => {
      console.log("Countries: ", res)

      dispatch({ type: types.FETCH_COUNTRIES, payload: res })
    }
  )
}