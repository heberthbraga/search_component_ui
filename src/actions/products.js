import types from '../types'

import ProductsApiService from '../api/products'

export const search = (searchTerm, aggregation, currentPage='1') => dispatch => {
  console.log("Start searching term, aggregation and page", searchTerm + ', ' + JSON.stringify(aggregation) + ', ' + currentPage)

  dispatch({ type: types.SEARCH_TERM, payload: searchTerm })

  ProductsApiService.search(searchTerm, aggregation, currentPage).then(
    res => {
      console.log("Search response: ", res)

      dispatch({ type: types.SEARCH_PRODUCTS, payload: res })
    }
  )
}

export const handleSortOption = (option) => dispatch => ( dispatch({ type: types.SORT_PRODUCTS, payload: option }) )

export const initAggregation = (aggregation) => dispatch => ( dispatch({ type: types.INIT_AGGREGATION, payload: aggregation }) )