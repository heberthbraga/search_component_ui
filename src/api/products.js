import { post, get } from './call'

const ProductsApiService = {
  search: (searchTerm, aggregation, page) =>
    post('products/search', { text: searchTerm, aggregation: aggregation, page }),

  list: () =>
    get('products')
}

export default ProductsApiService