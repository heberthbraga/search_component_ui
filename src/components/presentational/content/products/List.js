import React from 'react'

import styled from 'styled-components'

import Layout from '../../application/Layout'
import CircleSpinner from '../../shared/CircleSpinner'

const listItems = (products) => {
  return products.map((p) => {
    return (
      <GridItem key={p.id}>
        <GridItemContent>
          <h3>{p.title}</h3>
          <p>${p.price}</p>
          <p>{p.description}</p>
        </GridItemContent>
      </GridItem>
    )
  })
}

const List = ({ products }) => {
  return (
    <Layout
      displayNav={true}
    >
      <Container>
        {!products ?
          <CircleSpinner />
        :
          <GridRow>
            {
              listItems(products)
            }
          </GridRow>
        }
      </Container>
    </Layout>
  )
}

const Container = styled.div`
  margin: 0 auto;
`

const GridRow = styled.ul`
  padding: 10px;
  display: grid;
  overflow: hidden;
  grid-template-columns: repeat(3, 1fr);
  grid-auto-rows: 1fr;
  grid-column-gap: 5px;
  grid-row-gap: 5px;
`

const GridItem = styled.li`
  background-color: #ccc;
  display: flex;
  padding: 0.5em;
  margin-bottom: 20px;
`

const GridItemContent = styled.div`
  width: 100%;
`

export default List