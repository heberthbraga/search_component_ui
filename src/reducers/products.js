import types from '../types'

const initialState={
  list: [],
  searchTerm: ''
}

const reducer = (state=initialState, action) => {

  if (action.type === types.SEARCH_PRODUCTS) {
    return {
      ...state,
      list: action.payload
    }
  }

  if (action.type === types.SEARCH_TERM) {
    return {
      ...state,
      searchTerm: action.payload
    }
  }

  if (action.type === types.REDIRECT_TO_HOME) {
    const { routing } = state
    state = { routing } 
  }

  return state
}

export default reducer
