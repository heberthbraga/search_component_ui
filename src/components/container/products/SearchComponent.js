import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { push } from 'connected-react-router'

import * as actions from '../../../actions/products'

import SearchFrom from '../../presentational/forms/SearchForm'

import { sortOptions } from '../../../helpers'

class SearchComponent extends Component {
  constructor(props) {
    super(props)

    this.handleSubmit = this.handleSubmit.bind(this)
  }

  componentWillMount() {
    this.props.handleSortOption(sortOptions.default.value)
  }

  handleSubmit = (e) => {
    e.preventDefault()

    const { search, changePage, aggregation } = this.props

    let searchTerm = e.target.elements.term.value ? e.target.elements.term.value : ''

    console.log(aggregation)

    search(searchTerm, aggregation)
    changePage()
  }

  render() {
    const { searchTerm } = this.props

    return (
      <SearchFrom 
        handleSubmit={this.handleSubmit}
        searchTerm={searchTerm}
      />
    )
  }
}

const mapStateToProps = state => ({
  products: state.products.list,
  searchTerm: state.products.searchTerm,
  aggregation: state.products.aggregation
})

const mapDispatchToProps = dispatch => bindActionCreators({
  changePage: () => push('/search'),
  handleSortOption: (option) => actions.handleSortOption(option),
  search: (searchTerm, aggregation) => actions.search(searchTerm, aggregation)
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(SearchComponent)