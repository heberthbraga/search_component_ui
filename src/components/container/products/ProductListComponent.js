import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { Redirects } from '../../../routes'
import * as productActions from '../../../actions/products'

import List from '../../presentational/content/products/List'

class ProductListComponent extends Component {
  constructor(props) {
    super(props)
  }

  onPageChange = (action) => (targetPage) => {
    const { changePage, search, searchTerm, aggregation } = this.props

    if (action === 'search' && targetPage) {
      search(searchTerm, aggregation, targetPage)
      changePage()
    }
  }

  render() {
    const { products, action, searchTerm } = this.props
    
    return (
      <List 
        products={products}
        action={action}
        searchTerm={searchTerm}
        onPageChange={this.onPageChange(action)}
      />
    )
  }
}

const mapStateToProps = state => ({
  products: state.products.list,
  searchTerm: state.products.searchTerm,
  aggregation: state.products.aggregation
})

const mapDispatchToProps = dispatch => bindActionCreators({
  changePage: () => Redirects.gotToSearch(),
  search: (searchTerm, aggregation, currentPage) => productActions.search(searchTerm, aggregation, currentPage)
}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(ProductListComponent)