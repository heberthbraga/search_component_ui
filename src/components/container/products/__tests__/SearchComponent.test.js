import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'

import SearchComponent from  '../SearchComponent'

import products from '../../../../reducers/products'

import * as productActions from '../../../../actions/products'

jest.mock('../../../../actions/products.js')

const mockStore = configureStore([])

describe('SearchComponent', () => {
    let store
    let component

    beforeEach(()=> {
      store = mockStore({
        products
      })

      component = renderer.create(
        <Provider store={store}>
          <SearchComponent />
        </Provider>
      )
    })

    it('should render the container correctly', () => {
      expect(component.toJSON()).toMatchSnapshot()
    })
})