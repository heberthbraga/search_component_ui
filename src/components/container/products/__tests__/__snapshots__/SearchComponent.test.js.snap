// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent should render 1`] = `
<form
  className="sc-fzXfNN laLdhK"
  onSubmit={[Function]}
>
  <input
    className="sc-fzXfNO dnqrIa"
    name="term"
    type="text"
  />
  <button
    className="sc-fzXfNL sc-fzXfNM kLJjdL"
  >
    <span
      className="sc-fzXfNP eBhAfH"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-search fa-w-16 "
        data-icon="search"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </span>
  </button>
</form>
`;

exports[`SearchComponent should render the component correctly 1`] = `
<form
  className="sc-fzXfNN laLdhK"
  onSubmit={[Function]}
>
  <input
    className="sc-fzXfNO dnqrIa"
    name="term"
    type="text"
  />
  <button
    className="sc-fzXfNL sc-fzXfNM kLJjdL"
  >
    <span
      className="sc-fzXfNP eBhAfH"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-search fa-w-16 "
        data-icon="search"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </span>
  </button>
</form>
`;

exports[`SearchComponent should render the container correctly 1`] = `
<form
  className="sc-fzXfNN laLdhK"
  onSubmit={[Function]}
>
  <input
    className="sc-fzXfNO dnqrIa"
    name="term"
    type="text"
  />
  <button
    className="sc-fzXfNL sc-fzXfNM kLJjdL"
  >
    <span
      className="sc-fzXfNP eBhAfH"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-search fa-w-16 "
        data-icon="search"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </span>
  </button>
</form>
`;
